### PR TITLE
test LitAPI decode-encode methods

### PR DESCRIPTION
**Pull request by aniketmaurya**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## ⚠️ How does this PR impact the user? ⚠️
Describe (in plain English, not technical Jargon) how this improves the user experience. If you can't tie it back to a real tangible, user goal or describe it in plain english, it's a hint that this is likely not needed and is probably an "engineering nit". 

### GOOD:
"As a user, I need to serve models faster. This PR focuses on enabling speed gains by using GPUs"

### BAD:
"This PR enables GPUs". 
This is bad because the *user problem* is not clear... instead it just jumps to the solution without any context. 

PRs without this will not be merged.

---- 

## What does this PR do?

Fixes # (issue).

Ensure that the decode_request and encode_response methods are thoroughly tested across various scenarios and inputs to enhance robustness.

**Increases the test coverage by 2%**.

```
@@         Coverage Diff         @@
##           main   #244   +/-   ##
===================================
+ Coverage    82%    84%   +2%     
===================================
  Files        13     13           
  Lines      1055   1054    -1     
===================================
+ Hits        861    881   +20     
+ Misses      194    173   -21     
```

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/litserve/specs/openai.py [hunk 1]
```diff
@@ -311,16 +311,14 @@ def _encode_response(self, output: Union[Dict[str, str], List[Dict[str, str]]])
         elif isinstance(output, dict) and "content" in output:
             message = output.copy()
             message.update(role="assistant")
-        elif self.validate_chat_message(output):
-            message = output
```
Comment (Andrei-Aksionov) [inline src/litserve/specs/openai.py:315]: Why did you remove it?

I see that the check above partially matches `validate_chat_message`, perhaps that why?
But in the check above you overwrite `role` anyway, even if it was already in the dict.

So maybe keep it, but move above  `elif isinstance(output, dict) and "content" in output:`

**Review by Andrei-Aksionov — APPROVED** [review]

File: tests/test_litapi.py [hunk 2]
```diff
@@ -79,3 +95,90 @@ def test_batch_unbatch_stream():
...
+    api._sanitize(max_batch_size=1, spec=ls.OpenAISpec())
+    request = ChatCompletionRequest(messages=[{"role": "system", "content": "Hello"}])
+    decoded_request = api.decode_request(request)
+    assert decoded_request[0]["content"] == "Hello", "Decode request should return the input message"
+
+
+def test_decode_request_with_openai_spec_wrong_request():
+    api = ls.examples.TestAPI()
+    api._sanitize(max_batch_size=1, spec=ls.OpenAISpec())
+    with pytest.raises(AttributeError, match="object has no attribute 'messages'"):
```
Comment (lantiga) [inline tests/test_litapi.py:117]: let's extend the failure modes here

File: tests/test_litapi.py [hunk 3]
```diff
@@ -79,3 +95,90 @@ def test_batch_unbatch_stream():
...
+        api.decode_request({"input": "Hello"})
+
+
+def test_encode_response():
+    response = 4.0
+    api = ls.examples.SimpleLitAPI()
+    assert api.encode_response(response) == {"output": 4.0}, 'Encode response returns encoded output {"output": 4.0}'
```
Comment (lantiga) [inline tests/test_litapi.py:124]: let's extend encoding with more inputs types

File: tests/test_litapi.py [hunk 4]
```diff
@@ -79,3 +95,90 @@ def test_batch_unbatch_stream():
...
+    generated_tokens = []
+    for output in api.encode_response(response):
+        generated_tokens.append(output["content"])
+    assert generated_tokens == response, f"Encode response should return the generated tokens {response}"
+
+
+def test_encode_response_with_openai_spec_invalid_input():
+    api = ls.examples.TestAPI()
+    api._sanitize(max_batch_size=1, spec=ls.OpenAISpec())
+    response = 10
```
Comment (lantiga) [inline tests/test_litapi.py:172]: let's extend failure modes

**Review by lantiga — COMMENTED** [review]
Looks good, but we should extend failure modes

File: tests/test_litapi.py (hunk 3 above)
Comment (aniketmaurya) [reply tests/test_litapi.py:124]: It is implemented by SimpleLitAPI so for more input types, we would be testing the SimpleLitAPI class then

File: tests/test_litapi.py (hunk 2 above)
Comment (aniketmaurya) [reply tests/test_litapi.py:117]: default decode_request is Pydantic typed so it handles anything other than a ChatCompletionRequest object with the same AttributeError.
